### DISCUSSION TITLE
Working with drafts

### DISCUSSION THREAD

**Discussion by marcuslindblom** [ravendb/ravendb#19316] [Q&A] [answered]
In our current solution we have stored our draft data in a separate document with the same id but with `/draft` at the end.
We also use an index to load documents based on their url so a query would look like this
```js
const page =  await session
  .query({ indexName: 'Content/ByUrl' })
  .whereEquals('url', '/my-awesome-post')
  .single();
```
using this Index 👇🏻 
```js
map('Posts', c => {
              let collection = c['@metadata']['@collection']
              let template = load('templates/' + collection, 'templates')
              if (!c.archived && template && !template.disableURL) {
                let id = Id(c)
                let doc = c
                let visited = {}
                let slugs = []
                let origin = c.origin.id
                let displayName = c.displayName
                let publishedDate = c.publishedDate
                let deleted = c.deleted
                let createdBy = c.createdBy
                let createdAt = c.createdAt
                do {
                  if (c.slug) {
                      slugs.unshift(c.slug)
                  }
                  if (visited[c.origin.id])
                    break
                  visited[c.origin.id] = true
                  c = load(c.origin.id, c.origin.collection)
                } while(c);
                return {
                  id: id,
                  displayName: displayName,
                  url: slugs.shift() + slugs.join('/'),
                  origin: origin,
                  collection: collection,
                  publishedDate: publishedDate,
                  createdBy: createdBy,
                  createdAt: createdAt,
                  deleted: deleted,
                };
              }
          });
```
This works great but often I find my self wanting to do something like this.

```
const page = await session
  .query({indexName: 'Content/ByUrl')
  .whereEquals('url', '/my-awesome-post')
  .whereEquals('_status', 'draft');
```
But then I need to index all drafts with the same url as the once published and that can certainly some problems.

So, with that said, what do you recommend when working with drafts and conditional loading? Different documents, same document with the new value in the same document like `name: 'Marcus', _name: 'Marcus draft value'`, or third option?

**Comment by ayende** [comment, marked as answer]
I'm afraid that there isn't enough information here to understand the problem.

You have a document with `@id = 'posts/1234` and  `{ "url": "/my-awesome-post" }` (ignoring the hierarchy for now). 
You also have _another_ one with `@id = 'posts/1234/draft` and  `{ "url": "/my-awesome-post" }` , right?

And you:

* Sometimes want to load the draft
* Sometimes want to load the real deal using `single` 

If you'll index the draft, you are worried that `single` will break?

If you want to load draft by url, can you not do:

```
url: slugs.shift() + slugs.join('/') + (c._status == 'draft' ? "?draft=true" : "")
```

And the when you query you'll have:

```
const page =  await session
  .query({ indexName: 'Content/ByUrl' })
  .whereEquals('url', '/my-awesome-post')
  .single();
```

versus:

```
const page =  await session
  .query({ indexName: 'Content/ByUrl' })
  .whereEquals('url', '/my-awesome-post?draft=true')
  .single();
```
